### ISSUE TITLE
Add class_weight to clone detection (binary problem)

### ISSUE THREAD

**Issue by Muradean** [microsoft/CodeBERT#64]
I have been using the clone detection folder to perform a binary classification task. However my classes are not balanced, so I followed some posts and changed the forward  function in model.py to have class weights:

Before:

          
      def forward(self, input_ids=None,labels=None): 
          logits=self.encoder(input_ids,attention_mask=input_ids.ne(1))[0]
          prob=torch.softmax(logits,-1)
          if labels is not None:
  
              weights = torch.tensor([1., 2.]).cuda()
  
              loss_fct = nn.CrossEntropyLoss(ignore_index=-1,weight=weights)
              loss = loss_fct(logits,labels)
              return loss,prob
          else:
              return prob
        
After:

      def forward(self, input_ids=None,labels=None): 
          logits=self.encoder(input_ids,attention_mask=input_ids.ne(1))[0]
          prob=torch.softmax(logits,-1)
          if labels is not None:
              weights = [0.001,1000000000000]

              class_weights = torch.FloatTensor(weights).cuda()
              oss_fct = nn.CrossEntropyLoss(ignore_index=-1,weight=class_weights)            

            loss = loss_fct(logits,labels)
            return loss,prob
        else:
            return prob
      
          

        

However, I have tried different weights, and different configurations (giving more weight to each class respectively) But the model just predicts always the same class... Does anyone have any idea of what might be going on ?

Thanks a lot

**Comment by guoday** [comment]
You can print the probability of prediction. If the probabilities of all test examples are same, you need to check whether the learning rate is too large. Otherwise, the training of model is ok, but the probability of other class is hard to be large than 0.5 so that the model always predicts the same class. You need to sample some negative for training to keep balance.

**Comment by Muradean** [comment]
Thanks for replying, I realized what was going on, it had to do with the batch_size, because despite greatly increasing the weight of one of the classes, my batch_size (10) was not big enough. As a result, almost every training iteration only had instances from the majority class. The solution was to increase the batch size to 40. Now all instances are predicted as the minority class, I guess I'd just need to tweak the parameters a little bit, which should be enough.

Batch size = 10:
![batch_size_10](https://user-images.githubusercontent.com/14300941/130705361-69b46b01-1e7a-4506-890b-c576d1f17b87.png)

Batch size = 40:

![last_probabilities_batch_size_40](https://user-images.githubusercontent.com/14300941/130705374-2f375611-3cb7-430a-8a9d-9c280685646e.png)
